Packages/io.chainsafe.web3-unity/Runtime/Scripts/WalletConnect/Dialog/ConnectionDialogChainSafe.cs

```diff
@@ -0,0 +1,82 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using UnityEngine;
+using UnityEngine.UI;
+using ZXing.QrCode;
+
+namespace ChainSafe.Gaming.WalletConnect.Dialog
+{
+    public class ConnectionDialogChainSafe : ConnectionDialogBase
+    {
+        public new Animation animation;
+        public string showAnimationClipName;
+        public string hideAnimationClipName;
+        public RawImage qrCodeImage;
+        public Vector2Int qrCodeSize = new(512, 512);
+        public int qrCodePadding = 20;
+        public GameObject qrCodePanel;
+        public GameObject localWalletButtonsPanel;
+        public LocalWalletButton localWalletButtonPrefab;
+        public RectTransform localWalletButtonsContainer;
+        public GameObject singleButtonForLocalWalletsPanel;
+        public Button singleButtonForLocalWallets;
+
+        private LocalWalletButton[] loadedLocalWalletButtons;
+
+        private void Awake()
+        {
+            singleButtonForLocalWallets.onClick.AddListener(OpenLocalWalletOsManaged);
+        }
+
+        protected override void PlayShowDialog() => animation.Play(showAnimationClipName);
+        protected override void PlayHideDialog() => animation.Play(hideAnimationClipName);
+
+        protected override void SetRedirectOptionsVisible(bool visible) => localWalletButtonsPanel.SetActive(visible);
+        protected override void SetQrCodeElementVisible(bool visible) => qrCodePanel.SetActive(visible);
+
+        protected override void SetSingleButtonForRedirectVisible(bool visible) =>
+            singleButtonForLocalWalletsPanel.SetActive(visible);
+
+        protected override void SpawnRedirectOptions(List<WalletOptionConfig> supportedWallets)
+        {
+            loadedLocalWalletButtons = supportedWallets
+                .Select(w =>
+                {
+                    var button = Instantiate(localWalletButtonPrefab, localWalletButtonsContainer);
+                    button.Set(w.Data, w.OnClick);
+                    return button;
+                })
+                .ToArray();
+        }
+
+        protected override void CreateQrCodeElement(QrCodeBuilder builder)
+        {
+            qrCodeImage.texture = builder.GenerateQrCode(new QrCodeEncodingOptions
+                {
+                    Width = qrCodeSize.x,
+                    Height = qrCodeSize.y,
+                    Margin = qrCodePadding
+                }
+            );
+        }
+
+        protected override void ClearDynamicElements()
+        {
+            if (loadedLocalWalletButtons != null)
+            {
+                foreach (var localWalletButton in loadedLocalWalletButtons)
+                {
+                    Destroy(localWalletButton.gameObject);
+                }
+
+                loadedLocalWalletButtons = null;
+            }
+        }
+
+        public void Close()
+        {
+            OnException(new Exception("User closed the connection dialog."));
```

Comment:
This is how we tell the system that the connection process was canceled. This is basically `Promise.Reject()`. It works because we're awaiting the task upper in the stack. Awaiting method then handles this accordingly. You can look into `WalletConnectProvider.ConnectSession()` to see how it's being handled.

Reply:
I saw how the `TaskCompletionSource` works for this but why are throwing an exception when a user explicitly cancels the connection? shouldn't we just dispose of any initialized connection objects/prepare for a new connection request?

Reply:
I don't think I understand what you're suggesting.

Reply:
Closing the connection dialog UI shouldn't throw an exception instead it should dispose of initialized objects and prepare for another connection, I've realized you mentioned this on the PR description under known issues
> You can't log in again after you've closed the connection dialog (should be fixed as soon as possible).
This should be fixed in this PR